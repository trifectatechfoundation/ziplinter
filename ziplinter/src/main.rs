use rc_zip_sync::ReadZip;
use std::ops::Deref;

fn main() {
    let mut args = std::env::args();
    let _ = args.next();
    let path = args.next().unwrap();

    let file = std::fs::File::open(path).unwrap();
    let archive = file.read_zip().unwrap();

    eprintln!(
        "{}",
        serde_json::to_string_pretty(&archive.deref()).unwrap()
    )
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;
    use std::{error::Error, path::Path};

    fn process_zip_file(zip_path: &Path) -> Result<serde_json::Value, Box<dyn Error>> {
        let file = std::fs::File::open(zip_path).unwrap();
        let archive = file.read_zip();

        Ok(serde_json::to_value(archive?.deref())?)
    }

    #[test]
    fn snapshot_zip_files() {
        let fixtures_dir = std::env::current_dir()
            .unwrap()
            .join("../testdata")
            .canonicalize()
            .unwrap();

        println!("fixtures_dir: {}", fixtures_dir.display());

        for entry in std::fs::read_dir(fixtures_dir).expect("Failed to read fixtures directory") {
            let entry = entry.expect("Failed to read entry");
            let path = entry.path();

            if path
                .extension()
                .is_some_and(|e| e.eq_ignore_ascii_case("zip"))
            {
                let Ok(result) = process_zip_file(&path) else {
                    continue;
                };
                println!(
                    "current file: {}",
                    Path::new(path.file_name().unwrap()).display()
                );
                assert_json_snapshot!(
                    format!("{}", Path::new(path.file_name().unwrap()).display()),
                    result
                );
            }
        }
    }
}
