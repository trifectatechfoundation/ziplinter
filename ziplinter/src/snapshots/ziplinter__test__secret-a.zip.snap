---
source: ziplinter/src/main.rs
expression: result
---
{
  "comment": "",
  "contents": [
    {
      "central": {
        "aex": {
          "compression_method": 8,
          "mode": 3,
          "vendor_id": 17729,
          "version_number": 1
        },
        "comment": "",
        "compressed_size": 34,
        "crc32": 1311606447,
        "creator_version": {
          "host_system": "Unix",
          "version": 20
        },
        "disk_nbr_start": 0,
        "external_attrs": 2175008768,
        "extra": [
          85,
          84,
          13,
          0,
          7,
          12,
          162,
          202,
          103,
          12,
          162,
          202,
          103,
          12,
          162,
          202,
          103,
          117,
          120,
          11,
          0,
          1,
          4,
          251,
          3,
          0,
          0,
          4,
          254,
          3,
          0,
          0,
          1,
          153,
          7,
          0,
          1,
          0,
          65,
          69,
          3,
          8,
          0
        ],
        "flags": 9,
        "header_offset": 0,
        "internal_attrs": 0,
        "method": "Aex",
        "mode": 420,
        "modified": "2025-03-07T07:36:44Z",
        "name": "a.txt",
        "reader_version": {
          "host_system": "MsDos",
          "version": 20
        },
        "uncompressed_size": 21
      },
      "local": {
        "accessed": null,
        "aex": {
          "compression_method": 8,
          "mode": 3,
          "vendor_id": 17729,
          "version_number": 1
        },
        "aex_data": {
          "authentication_code": [
            137,
            221,
            245,
            0,
            163,
            143,
            18,
            105,
            215,
            13
          ],
          "password_verification_value": [
            18,
            132
          ],
          "salt_value": [
            17,
            177,
            73,
            166,
            170,
            40,
            101,
            191,
            211,
            6,
            21,
            32,
            146,
            220,
            178,
            2
          ]
        },
        "compressed_size": 0,
        "crc32": 0,
        "created": null,
        "extra": [
          85,
          84,
          13,
          0,
          7,
          12,
          162,
          202,
          103,
          12,
          162,
          202,
          103,
          12,
          162,
          202,
          103,
          117,
          120,
          11,
          0,
          1,
          4,
          251,
          3,
          0,
          0,
          4,
          254,
          3,
          0,
          0,
          1,
          153,
          7,
          0,
          1,
          0,
          65,
          69,
          3,
          8,
          0
        ],
        "flags": 9,
        "gid": 1019,
        "header_offset": 0,
        "method": "Aex",
        "method_specific": "None",
        "mode": 0,
        "modified": "2025-03-07T07:36:44Z",
        "name": "a.txt",
        "reader_version": {
          "host_system": "MsDos",
          "version": 20
        },
        "uid": 1019,
        "uncompressed_size": 21
      }
    }
  ],
  "encoding": "Utf8",
  "eocd": {
    "dir": {
      "inner": {
        "dir_disk_nbr": 0,
        "dir_records_this_disk": 1,
        "directory_offset": 128,
        "directory_records": 1,
        "directory_size": 94,
        "disk_nbr": 0
      },
      "offset": 222
    },
    "dir64": null,
    "global_offset": 0
  },
  "parsed_ranges": [
    {
      "contains": "end of central directory record",
      "end": 244,
      "start": 222
    },
    {
      "contains": "central directory header",
      "end": 222,
      "filename": "a.txt",
      "start": 128
    },
    {
      "contains": "local file header",
      "end": 78,
      "filename": "a.txt",
      "start": 0
    },
    {
      "contains": "file data",
      "end": 112,
      "filename": "a.txt",
      "start": 78
    },
    {
      "contains": "data descriptor",
      "end": 128,
      "filename": "a.txt",
      "start": 112
    }
  ],
  "size": 244
}
